### DISCUSSION TITLE
Monaco editor mobile support and web-terminal selection

### DISCUSSION THREAD

**Discussion by benalleng** [saving-satoshi/saving-satoshi#467] [General]
I know its after the fact but quite frankly we weren't focusing on mobile when we were designing the scripting challenge but after @ecurrencyhodler 's and others' QA session i think its clear that the monaco editor is a tough sell on mobile. Microsoft has stated no [official support](https://microsoft.github.io/monaco-editor/) for the monaco editor on mobile.

Large discussion about monaco on mobile [here](https://github.com/microsoft/monaco-editor/issues/246)
* The gist is no support seems planned

With our discussion in last-weeks call 5/16 about moving future coding challenges to repl's i wonder if we should consider the same for scripting-2 at the very least for mobile devices.

I know we spent a lot of time working on this! but with all future coding challenges planned for repl's I think there is some value in consistency.

**Comment by ecurrencyhodler** [comment]
Thanks for raising this ben! During our call today, we decided we will find a longterm solution for using a coding environment inside our game. So we won't need to focus on this.

**Comment by afterburn** [comment]
After some internal discussions we think its best if we continue to use monaco editor. The unpleasantness @adamjonas experienced is mostly originating from our wasm python and js "compiler" and should no longer be issues once we move to an actual compiler. We just need to fix the issues @ecurrencyhodler reported and the issues reported by @benalleng:

Keyboard related issues:
- When typing the characters are not properly tracked. When I type in characters sometimes they lag behind in rendering by one character even though they are technically recognized as having been typed
- A focus issue when the editor loses focus you need to click somewhere else before clicking the editor to open again

Other issues around the editor but not directly related to Monaco
- The runner tab has a fixed position so it jumps in front of the editor when the virtual keyboard is displayed

**Comment by benalleng** [comment]
@afterburn I'm sure you have also been looking at some code-compiling options but i came across [Judge0](https://ce.judge0.com/)
I wonder if you think this is a good solution either via building it into a backend ourselves or going with their hosting options
[Plans](https://rapidapi.com/judge0-official/api/judge0-ce/pricing)
[Self-hosting docker guide](https://github.com/judge0/judge0/blob/master/CHANGELOG.md#deployment-procedure)

**Reply by benalleng** [reply]
the also have an IDE to play around in https://ide.judge0.com/

**Reply by afterburn** [reply]
One requirement we have is that the solution supports runtime evaluation, what I mean by that is that we should be able to write a loop, i.e.:
```py
import time

nonce = 0

while nonce < 20:
  print(nonce)
  time.sleep(1)
```

We should then be able to hook into stdout so every second we should be able to receive a message (whether it be via websocket, sse or polling doesn't matter) so we can log it out to the console/runner as the script is executing.

At a quick glance Judge0 seems to be submission based like many others (i.e. send request to submit code, send request to get result), which doesn't lend itself well to things like our hashing challenge.

**Reply by afterburn** [reply]
Their IDE seems to be using monaco judging by the context menu and network tab:

![image](https://github.com/saving-satoshi/saving-satoshi/assets/10757768/a1e85ffd-5fda-4796-8cb4-9460b3a284c3)

![image](https://github.com/saving-satoshi/saving-satoshi/assets/10757768/9064abf8-8389-4b9b-836c-66f780a99f7b)

**Reply by benalleng** [reply]
Nice idea to figure out what they are using as their frontend!
I see you ran a while loop, I also tested it, and it did seem to run the loop. I can understand why having a loop would be difficult in a classic submission based code compiler, why do you think it works here?
Do you think this is in the right direction as far as compiler options goes?

**Reply by benalleng** [reply]
I did try a loop more like our challenge and it did seem to work also! though it doesn't work past a certain point if it can't find it or the output is too large. I wonder if you think there are some additional limitations that we might run into.

![image](https://github.com/saving-satoshi/saving-satoshi/assets/108441023/156d7318-b870-44fc-aaf9-70e3d92ee61c)

**Reply by afterburn** [reply]
Yes, it runs the loop but the if you check your network tab you can see that the full stdout will be sent to the client only _after_ the script finishes executing:

https://github.com/saving-satoshi/saving-satoshi/assets/10757768/e54d6573-a43e-42a2-8afd-e142c60c77a7

What we need instead is a solution that'll let's us hook into stdout _while_ the script is executing, so more like this:

https://github.com/saving-satoshi/saving-satoshi/assets/10757768/4068114a-e651-4fc1-90e4-3f2100f2d51e

Notice the difference between the two when I add `time.sleep(1)`, Judge0 will simply stop executing after the alotted time passes and will not return any result, while the api in the video below works because it pipes stdout to the client as opposed to only sending the script output after execution finishes.

The reason we need this is because in our scripting challenge we want to show each hash along with an evaluation message individually, with some delay before hashing the next nonce. We could theoretically fake a solution, but I think it'd be better if we can just hook into stdout, which is a more robust solution for future challenges.
